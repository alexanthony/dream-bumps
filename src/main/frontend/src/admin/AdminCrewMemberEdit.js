// @flow
import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import { Field, FieldArray, reduxForm, formValueSelector } from 'redux-form';
import { Card, CardTitle, CardText } from 'material-ui/Card';
import { TextField } from 'redux-form-material-ui';
import RaisedButton from 'material-ui/RaisedButton';

import { getCrewMembers } from '../rootReducer';
import { updateCrewMembers } from '../crews/actions';
import seatNumberToName from '../utils/seatNumberToName';
/**
 * AdminCrewDetailsEdit
 */
const renderMembers = ({ fields, current }) => (
  <div>
    {fields.map((member, index) => (
      <Field
        name={`${member}.name`}
        component={TextField}
        floatingLabelText={current ? seatNumberToName(current[index].seat) : ''}
        fullWidth
        key={index}
      />
    ))}
  </div>
);

renderMembers.propTypes = {
  fields: PropTypes.arrayOf(PropTypes.string),
  current: PropTypes.arrayOf(
    PropTypes.shape({
      seat: PropTypes.number,
      name: PropTypes.string
    })
  )
};

const AdminCrewMemberEdit = props => (
  <Card>
    <CardTitle title={`Crew Members - ${props.crewId}`} />
    <CardText>
      <form onSubmit={props.handleSubmit}>
        <FieldArray
          name="members"
          component={renderMembers}
          props={{ current: props.currentMembers }}
        />
        <RaisedButton primary type="submit" label="Save" />
      </form>
    </CardText>
  </Card>
);

AdminCrewMemberEdit.propTypes = {
  crewId: PropTypes.string,
  handleSubmit: PropTypes.func,
  currentMembers: PropTypes.arrayOf(
    PropTypes.shape({
      seat: PropTypes.number,
      name: PropTypes.string,
      id: PropTypes.number
    })
  ),
  initialValues: PropTypes.shape({
    members: PropTypes.arrayOf(
      PropTypes.shape({
        seat: PropTypes.number,
        name: PropTypes.string,
        id: PropTypes.number
      })
    )
  })
};

const selector = formValueSelector('edit-crew-members');

export const mapStateToProps = (
  state: Object,
  { crewId }: { crewId: string }
) => ({
  initialValues: { members: getCrewMembers(state)(parseInt(crewId, 10)) },
  currentMembers: selector(state, 'members')
});

export { AdminCrewMemberEdit };

export default connect(mapStateToProps, { onSubmit: updateCrewMembers })(
  reduxForm({ form: 'edit-crew-members', enableReinitialize: true })(
    AdminCrewMemberEdit
  )
);
