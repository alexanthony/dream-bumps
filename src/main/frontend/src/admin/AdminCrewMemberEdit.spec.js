import React from 'react';
import { shallow } from 'enzyme';
import { AdminCrewMemberEdit } from './AdminCrewMemberEdit';
import toJson from 'enzyme-to-json';

test('Renders properly', () => {
  const submit = jest.fn();
  const fetchCrewMembers = jest.fn();
  const component = shallow(
    <AdminCrewMemberEdit
      crewId="19"
      handleSubmit={submit}
      currentMembers={[{name: "Member1", seat: 1}, {name: "Member 2", seat: 2}, {name: "Member 3", seat: 3}]}
      initialValues={{members: [{name: "OldName1", seat: 1}, {name: "Old Name 2", seat: 2}, {name: "Member 3", seat: 3}]}}
      fetchCrewMembers={fetchCrewMembers}
    />
  );
  expect(component).toBeDefined();
  expect(toJson(component)).toMatchSnapshot()
});


// AdminCrewMemberEdit.propTypes = {
//   crewId: PropTypes.string,
//   handleSubmit: PropTypes.func,
//   currentMembers: PropTypes.array,
//   initialValues: PropTypes.object,
//   fetchCrewMembers: PropTypes.func
// };
