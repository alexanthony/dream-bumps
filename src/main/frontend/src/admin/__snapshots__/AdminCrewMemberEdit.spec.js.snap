exports[`test Renders properly 1`] = `
<Card
  expandable={false}
  expanded={null}
  initiallyExpanded={false}>
  <CardTitle
    title="Crew Members - 19" />
  <CardText>
    <form
      onSubmit={[Function]}>
      <FieldArray
        component={[Function]}
        name="members"
        props={
          Object {
            "current": Array [
              Object {
                "name": "Member1",
                "seat": 1,
              },
              Object {
                "name": "Member 2",
                "seat": 2,
              },
              Object {
                "name": "Member 3",
                "seat": 3,
              },
            ],
          }
        } />
      <RaisedButton
        disabled={false}
        fullWidth={false}
        label="Save"
        labelPosition="after"
        primary={true}
        secondary={false}
        type="submit" />
    </form>
  </CardText>
</Card>
`;
